File: 0000.程序员面试金典/面试题17.21.直方图的水量/solution.go
func trap(height []int) int {
	sz := len(height)
	l2r, r2l := make([]int, sz), make([]int, sz)
	for i := 1; i < sz; i++ {
		l2r[i] = max(l2r[i-1], height[i-1])
	}
	for i := sz - 2; i >= 0; i-- {
		r2l[i] = max(r2l[i+1], height[i+1])
	}
	var ans int
	for i := 1; i < sz-1; i++ {
		diff := min(l2r[i], r2l[i]) - height[i]
		if diff > 0 {
			ans += diff
		}
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: 0000.程序员面试金典/面试题17.21.直方图的水量/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int sz=height.size();
        vector<int> l2r(sz),r2l(sz);
        for (int i=1;i<sz;i++) {
            l2r.at(i)=max(l2r.at(i-1),height.at(i-1));
        }
        for (int i=sz-2;i>=0;i--) {
            r2l.at(i)=max(r2l.at(i+1),height.at(i+1));
        }
        int ans=0;
        for (int i=1;i<sz-1;i++) {
            int diff=min(l2r.at(i),r2l.at(i))-height.at(i);
            ans+=diff>0 ? diff : 0;
        }
        return ans;
    }
};